test/GPv2TradeSimulator.test.ts

```diff
@@ -0,0 +1,233 @@
+    const vault = await waffle.deployMockContract(deployer, IVault.abi);
+
+    const GPv2Settlement = await ethers.getContractFactory(
+      "GPv2SettlementTestInterface",
+      deployer,
+    );
+    settlement = await GPv2Settlement.deploy(
+      authenticator.address,
+      vault.address,
+    );
+
+    const GPv2TradeSimulator = await ethers.getContractFactory(
+      "GPv2TradeSimulator",
+      deployer,
+    );
+    tradeSimulator = new TradeSimulator(
+      settlement,
+      await GPv2TradeSimulator.deploy(),
+    );
+
+    tokens = [
+      await waffle.deployMockContract(deployer, IERC20.abi),
+      await waffle.deployMockContract(deployer, IERC20.abi),
+    ];
+    router = await waffle.deployMockContract(deployer, IUniswapV2Router.abi);
+  });
+
+  describe("simulateTrade", () => {
+    it("computes account balance changes", async () => {
+      const sellAmount = ethers.utils.parseEther("42");
+      const sellRoundingError = 1;
+      const buyAmount = ethers.utils.parseEther("13.37");
+      const buyPositiveSlippage = 1001;
+
+      for (const receiver of [ethers.constants.AddressZero, receive.address]) {
+        const actualReceiver =
+          receiver === ethers.constants.AddressZero ? trader.address : receiver;
+        const swapArgs = [
+          sellAmount,
+          buyAmount,
+          [tokens[0].address, tokens[1].address],
+          settlement.address,
+          0xffffffff,
+        ];
+
+        await tokens[0].mock.balanceOf.withArgs(settlement.address).returns(0);
+        await tokens[1].mock.balanceOf.withArgs(settlement.address).returns(0);
+
+        await tokens[0].mock.balanceOf
+          .withArgs(trader.address)
+          .returns(sellAmount);
+        await tokens[1].mock.balanceOf.withArgs(actualReceiver).returns(0);
+
+        await tokens[0].mock.transferFrom
+          .withArgs(trader.address, settlement.address, sellAmount)
+          .returns(true);
+        await router.mock.swapExactTokensForTokens
+          .withArgs(...swapArgs)
+          .returns([sellAmount, buyAmount.add(buyPositiveSlippage)]);
+        await tokens[1].mock.transfer
+          .withArgs(actualReceiver, buyAmount)
+          .returns(true);
+
+        const { contractBalance, ownerBalance } =
+          await tradeSimulator.simulateTrade(
+            {
+              owner: trader.address,
+              sellToken: tokens[0].address,
+              buyToken: tokens[1].address,
+              receiver,
+              sellAmount,
+              buyAmount,
+            },
+            {
+              [InteractionStage.INTRA]: [
+                {
+                  target: router.address,
+                  callData: router.interface.encodeFunctionData(
+                    "swapExactTokensForTokens",
+                    swapArgs,
+                  ),
+                },
+              ],
+              // Use post-interactions to change mock configuration to simulate
+              // balance changes.
```

Comment:
I like the idea and the execution, but I wonder if it would be easier and more reliable to use real tokens instead of mocks, since in any case this is very close to an e2e test. I'm ok with how it is now still.

Reply:
I plan on adding an E2E test as well (in fact, I am writing one), PR was just getting too big.

Reply:
As promised: https://github.com/gnosis/gp-v2-contracts/pull/890